## 1. Dataset Overview

This notebook profiles the raw Diabetes 130-US Hospitals dataset before
any cleaning, transformation, database loading, or modeling.

The source dataset contains inpatient diabetic hospital encounters from
1999–2008. According to the project blueprint, the dataset grain is one
row per inpatient encounter.

The purpose of this notebook is to validate the structure of the raw
data and identify data-quality considerations that will influence the
PostgreSQL schema and later analytical pipeline.


In [ ]:
#import all the library 
import pandas as pd


,encounter_id,patient_nbr,race,gender,age,weight,admission_type_id,discharge_disposition_id,admission_source_id,time_in_hospital,...,citoglipton,insulin,glyburide-metformin,glipizide-metformin,glimepiride-pioglitazone,metformin-rosiglitazone,metformin-pioglitazone,change,diabetesMed,readmitted
0,2278392,8222157,Caucasian,Female,[0-10),?,6,25,1,1,...,No,No,No,No,No,No,No,No,No,NO
1,149190,55629189,Caucasian,Female,[10-20),?,1,1,7,3,...,No,Up,No,No,No,No,No,Ch,Yes,>30
2,64410,86047875,AfricanAmerican,Female,[20-30),?,1,1,7,2,...,No,No,No,No,No,No,No,No,Yes,NO
3,500364,82442376,Caucasian,Male,[30-40),?,1,1,7,2,...,No,Up,No,No,No,No,No,Ch,Yes,NO
4,16680,42519267,Caucasian,Male,[40-50),?,1,1,7,1,...,No,Steady,No,No,No,No,No,Ch,Yes,NO


In [11]:
#laod the Raw data 

df = pd.read_csv("../Data/raw/diabetic_data.csv")


df.head()

,encounter_id,patient_nbr,race,gender,age,weight,admission_type_id,discharge_disposition_id,admission_source_id,time_in_hospital,...,citoglipton,insulin,glyburide-metformin,glipizide-metformin,glimepiride-pioglitazone,metformin-rosiglitazone,metformin-pioglitazone,change,diabetesMed,readmitted
0,2278392,8222157,Caucasian,Female,[0-10),?,6,25,1,1,...,No,No,No,No,No,No,No,No,No,NO
1,149190,55629189,Caucasian,Female,[10-20),?,1,1,7,3,...,No,Up,No,No,No,No,No,Ch,Yes,>30
2,64410,86047875,AfricanAmerican,Female,[20-30),?,1,1,7,2,...,No,No,No,No,No,No,No,No,Yes,NO
3,500364,82442376,Caucasian,Male,[30-40),?,1,1,7,2,...,No,Up,No,No,No,No,No,Ch,Yes,NO
4,16680,42519267,Caucasian,Male,[40-50),?,1,1,7,1,...,No,Steady,No,No,No,No,No,Ch,Yes,NO


In [ ]:
#Inspecting Row and column of the Dataset
print(f"Raw data shape:{df.shape}")

Raw data shape:(101766, 50)


In [12]:
df.columns.to_list()

['encounter_id',
 'patient_nbr',
 'race',
 'gender',
 'age',
 'weight',
 'admission_type_id',
 'discharge_disposition_id',
 'admission_source_id',
 'time_in_hospital',
 'payer_code',
 'medical_specialty',
 'num_lab_procedures',
 'num_procedures',
 'num_medications',
 'number_outpatient',
 'number_emergency',
 'number_inpatient',
 'diag_1',
 'diag_2',
 'diag_3',
 'number_diagnoses',
 'max_glu_serum',
 'A1Cresult',
 'metformin',
 'repaglinide',
 'nateglinide',
 'chlorpropamide',
 'glimepiride',
 'acetohexamide',
 'glipizide',
 'glyburide',
 'tolbutamide',
 'pioglitazone',
 'rosiglitazone',
 'acarbose',
 'miglitol',
 'troglitazone',
 'tolazamide',
 'examide',
 'citoglipton',
 'insulin',
 'glyburide-metformin',
 'glipizide-metformin',
 'glimepiride-pioglitazone',
 'metformin-rosiglitazone',
 'metformin-pioglitazone',
 'change',
 'diabetesMed',
 'readmitted']

In [5]:
df.nunique()

encounter_id                101766
patient_nbr                  71518
race                             6
gender                           3
age                             10
weight                          10
admission_type_id                8
discharge_disposition_id        26
admission_source_id             17
time_in_hospital                14
payer_code                      18
medical_specialty               73
num_lab_procedures             118
num_procedures                   7
num_medications                 75
number_outpatient               39
number_emergency                33
number_inpatient                21
diag_1                         717
diag_2                         749
diag_3                         790
number_diagnoses                16
max_glu_serum                    3
A1Cresult                        3
metformin                        4
repaglinide                      4
nateglinide                      4
chlorpropamide                   4
glimepiride         

## 2. Dataset Grain

The expected grain of the source dataset is one row per inpatient
diabetic hospital encounter.

We will inspect the encounter and patient identifiers to understand
whether each row represents a unique encounter and how encounters
relate to patients.

In [13]:
print("Total rows:", len(df))
print("Unique encounters:", df["encounter_id"].nunique())
print("Unique patients:", df["patient_nbr"].nunique())

Total rows: 101766
Unique encounters: 101766
Unique patients: 71518


In [14]:
df["patient_nbr"].value_counts().head(10)

patient_nbr
88785891    40
43140906    28
1660293     23
23199021    23
88227540    23
23643405    22
84428613    22
92709351    21
90609804    20
89472402    20
Name: count, dtype: int64

In [17]:
df.isna().sum()

encounter_id                    0
patient_nbr                     0
race                            0
gender                          0
age                             0
weight                          0
admission_type_id               0
discharge_disposition_id        0
admission_source_id             0
time_in_hospital                0
payer_code                      0
medical_specialty               0
num_lab_procedures              0
num_procedures                  0
num_medications                 0
number_outpatient               0
number_emergency                0
number_inpatient                0
diag_1                          0
diag_2                          0
diag_3                          0
number_diagnoses                0
max_glu_serum               96420
A1Cresult                   84748
metformin                       0
repaglinide                     0
nateglinide                     0
chlorpropamide                  0
glimepiride                     0
acetohexamide 

## 3. Initial Data Quality: Placeholder Values

The raw dataset uses `"?"` as a placeholder for missing or unavailable
information in several columns.

Before cleaning the data, we quantify these placeholder values to
understand the extent of missingness and identify columns that require
special attention during the data-cleaning and database-design stages.

In [18]:
# Count how many '?' values exist in each column
missing_counts = (df == '?').sum()

# Show only columns that actually have at least one '?'
missing_counts = missing_counts[missing_counts > 0].sort_values(ascending=False)
print(missing_counts)

weight               98569
medical_specialty    49949
payer_code           40256
race                  2273
diag_3                1423
diag_2                 358
diag_1                  21
dtype: int64


In [19]:
question_mark_summary = pd.DataFrame({
    "missing_count": (df == "?").sum(),
    "missing_percentage": ((df == "?").sum() / len(df)) * 100
})

question_mark_summary = (
    question_mark_summary[question_mark_summary["missing_count"] > 0]
    .sort_values("missing_count", ascending=False)
)

question_mark_summary

,missing_count,missing_percentage
weight,98569,96.858479
medical_specialty,49949,49.082208
payer_code,40256,39.557416
race,2273,2.233555
diag_3,1423,1.398306
diag_2,358,0.351787
diag_1,21,0.020636


In [20]:
column_profile = pd.DataFrame({
    "data_type": df.dtypes,
    "unique_values": df.nunique(),
    "missing_values": df.isna().sum(),
    "question_marks": (df == "?").sum()
})

column_profile

,data_type,unique_values,missing_values,question_marks
encounter_id,int64,101766,0,0
patient_nbr,int64,71518,0,0
race,str,6,0,2273
gender,str,3,0,0
age,str,10,0,0
weight,str,10,0,98569
admission_type_id,int64,8,0,0
discharge_disposition_id,int64,26,0,0
admission_source_id,int64,17,0,0
time_in_hospital,int64,14,0,0


### Initial Observations

The raw dataset contains 101,766 encounter records and 71,518 unique
patient identifiers. The encounter identifier is unique across the
observed records, while multiple encounters can be associated with the
same patient identifier.

The dataset contains both standard missing values and `"?"` placeholder
values. The `"?"` placeholder is particularly prevalent in variables such
as `weight`, `medical_specialty`, and `payer_code`.

Several variables stored as integers appear to represent categorical
codes, including admission type, discharge disposition, and admission
source. Other variables are stored as strings but represent categorical
attributes, such as age, gender, medication indicators, and readmission
status.

The diagnosis fields contain a large number of distinct codes and will
require further investigation as part of the eventual database design.

Some variables have very low or no variation, including `examide` and
`citoglipton`, and will be reviewed later before making any retention or
exclusion decisions.

In [ ]:

#Raw 
print(f' gender : {df['gender'].value_counts()}')
print(f' race : {df["race"].value_counts()}')
print(f' age : {df['age'].value_counts()}')
print(f' readmitted : {df['readmitted'].value_counts()}')

 gender : gender
Female             54708
Male               47055
Unknown/Invalid        3
Name: count, dtype: int64
 race : race
Caucasian          76099
AfricanAmerican    19210
?                   2273
Hispanic            2037
Other               1506
Asian                641
Name: count, dtype: int64
 age : age
[70-80)     26068
[60-70)     22483
[50-60)     17256
[80-90)     17197
[40-50)      9685
[30-40)      3775
[90-100)     2793
[20-30)      1657
[10-20)       691
[0-10)        161
Name: count, dtype: int64
 readmitted : readmitted
NO     54864
>30    35545
<30    11357
Name: count, dtype: int64


In [24]:

#after droping na 
print(f' gender : {df['gender'].value_counts(dropna=False)}')
print(f' race : {df["race"].value_counts(dropna=False)}')
print(f' age : {df['age'].value_counts(dropna=False)}')
print(f' readmitted : {df['readmitted'].value_counts(dropna=False)}')

 gender : gender
Female             54708
Male               47055
Unknown/Invalid        3
Name: count, dtype: int64
 race : race
Caucasian          76099
AfricanAmerican    19210
?                   2273
Hispanic            2037
Other               1506
Asian                641
Name: count, dtype: int64
 age : age
[70-80)     26068
[60-70)     22483
[50-60)     17256
[80-90)     17197
[40-50)      9685
[30-40)      3775
[90-100)     2793
[20-30)      1657
[10-20)       691
[0-10)        161
Name: count, dtype: int64
 readmitted : readmitted
NO     54864
>30    35545
<30    11357
Name: count, dtype: int64


### Categorical Variable Observations

The categorical profiling shows that gender contains three observed
categories, including a small number of `Unknown/Invalid` records.
Race contains six observed categories, including `?` as a placeholder
for unavailable information.

Age is represented as ten predefined age bands rather than exact
patient ages. The largest groups are `[70-80)` and `[60-70)`.

The raw readmission outcome contains three categories: `NO`, `>30`, and
`<30`. The `<30` category contains 11,357 encounters, while the
remaining categories contain 90,409 encounters. This indicates that
30-day readmission is a minority outcome and will require an
imbalance-aware modeling strategy later.

No categorical values are transformed or removed at this profiling
stage. Data-cleaning decisions will be made after the complete raw
dataset has been profiled.

In [25]:
df["admission_type_id"].value_counts().sort_index()

admission_type_id
1    53990
2    18480
3    18869
4       10
5     4785
6     5291
7       21
8      320
Name: count, dtype: int64

## 5. Coded Variables

Several variables in the raw dataset are represented using integer
codes rather than descriptive labels. The dataset documentation provides
the corresponding meaning for these codes.

The coded variables include:

- `admission_type_id`
- `discharge_disposition_id`
- `admission_source_id`

The documentation-based mappings will be retained as reference metadata.
The raw numeric codes will not be arbitrarily interpreted as continuous
numeric measurements.

In [26]:
df["admission_type_id"].value_counts().sort_index()

admission_type_id
1    53990
2    18480
3    18869
4       10
5     4785
6     5291
7       21
8      320
Name: count, dtype: int64

In [27]:
df["discharge_disposition_id"].value_counts().sort_index()

discharge_disposition_id
1     60234
2      2128
3     13954
4       815
5      1184
6     12902
7       623
8       108
9        21
10        6
11     1642
12        3
13      399
14      372
15       63
16       11
17       14
18     3691
19        8
20        2
22     1993
23      412
24       48
25      989
27        5
28      139
Name: count, dtype: int64

In [28]:
df["admission_source_id"].value_counts().sort_index()

admission_source_id
1     29565
2      1104
3       187
4      3187
5       855
6      2264
7     57494
8        16
9       125
10        8
11        2
13        1
14        2
17     6781
20      161
22       12
25        2
Name: count, dtype: int64

### Observations

The admission, discharge, and admission-source variables are represented
as documented categorical codes rather than continuous numerical
measurements.

The observed distributions are highly uneven. A small number of codes
account for most encounters, while several codes occur only rarely.

The code values are retained in their raw form during profiling. Their
documented meanings will be used as reference information, while any
mapping, grouping, or treatment of rare categories will be decided
during the data-cleaning and feature-engineering stages.

These variables are potentially important encounter-level attributes
because admission and discharge characteristics are part of the project's
defined analytical and modeling scope.

In [ ]:
#the identifiers and patient relationship
encounters_per_patient = df["patient_nbr"].value_counts()

encounters_per_patient.describe()

count    71518.000000
mean         1.422942
std          1.090740
min          1.000000
25%          1.000000
50%          1.000000
75%          1.000000
max         40.000000
Name: count, dtype: float64

In [30]:
encounters_per_patient.value_counts().sort_index()

count
1     54745
2     10434
3      3328
4      1421
5       717
6       346
7       207
8       111
9        70
10       42
11       20
12       19
13       14
14        5
15        9
16        4
17        3
18        6
19        3
20        6
21        1
22        2
23        3
28        1
40        1
Name: count, dtype: int64

In [32]:


encounters_per_patient.head(20)

patient_nbr
88785891    40
43140906    28
1660293     23
23199021    23
88227540    23
23643405    22
84428613    22
92709351    21
90609804    20
89472402    20
23398488    20
37096866    20
88789707    20
29903877    20
88479036    19
88681950    19
97391007    19
3481272     18
3401055     18
24011577    18
Name: count, dtype: int64

### Patient–Encounter Relationship

The dataset contains 101,766 encounters associated with 71,518 unique
patient identifiers. This indicates that the source data is encounter-
level rather than patient-level.

Most patients have a single recorded encounter, while a subset has
multiple encounters. The maximum observed number of encounters for a
single patient identifier is 40.

This supports a one-to-many relationship between patient identifiers
and encounters for the purpose of database design. However, the repeated
encounters should not be interpreted as a complete longitudinal patient
history because the source dataset does not provide sufficient
information to reconstruct one.

In [33]:
patient_consistency = df.groupby("patient_nbr").agg(
    race_unique=("race", "nunique"),
    gender_unique=("gender", "nunique"),
    age_unique=("age", "nunique")
)

patient_consistency.describe()

,race_unique,gender_unique,age_unique
count,71518.000000,71518.000000,71518.000000
mean,1.003524,1.000042,1.022008
std,0.059959,0.006477,0.146712
min,1.000000,1.000000,1.000000
25%,1.000000,1.000000,1.000000
50%,1.000000,1.000000,1.000000
75%,1.000000,1.000000,1.000000
max,3.000000,2.000000,2.000000


In [34]:
patient_consistency[
    (patient_consistency["race_unique"] > 1) |
    (patient_consistency["gender_unique"] > 1) |
    (patient_consistency["age_unique"] > 1)
].head(20)

,race_unique,gender_unique,age_unique
patient_nbr,,,
1152,1,1,2
5220,1,1,2
30735,1,1,2
36810,1,1,2
40158,1,1,2
78435,1,1,2
108477,1,1,2
126783,1,1,2
133443,1,1,2


In [35]:
attribute_consistency_summary = {
    "Race varies": (patient_consistency["race_unique"] > 1).sum(),
    "Gender varies": (patient_consistency["gender_unique"] > 1).sum(),
    "Age varies": (patient_consistency["age_unique"] > 1).sum()
}

attribute_consistency_summary

{'Race varies': np.int64(249),
 'Gender varies': np.int64(3),
 'Age varies': np.int64(1574)}

In [36]:
attribute_consistency_pct = {
    "Race varies (%)": (patient_consistency["race_unique"] > 1).mean() * 100,
    "Gender varies (%)": (patient_consistency["gender_unique"] > 1).mean() * 100,
    "Age varies (%)": (patient_consistency["age_unique"] > 1).mean() * 100
}

attribute_consistency_pct

{'Race varies (%)': np.float64(0.34816409854861713),
 'Gender varies (%)': np.float64(0.004194748175284544),
 'Age varies (%)': np.float64(2.200844542632624)}

### Patient Attribute Consistency

Among the 71,518 unique patient identifiers, race is consistent across
repeated encounters for almost all patients. Only 249 patients (0.35%)
have more than one observed race value.

Gender is highly consistent, with only 3 patients (0.004%) showing more
than one observed gender value.

Age band shows more variation: 1,574 patients (2.20%) have more than one
observed age band across their encounters. Unlike race and gender, age
band can naturally change between encounters as patients age.

These results show that repeated encounter records generally contain
consistent demographic information, while age band can vary over time.
No values are modified at this profiling stage.

In [37]:
numeric_columns = df.select_dtypes(include="number").columns.tolist()

numeric_columns

['encounter_id',
 'patient_nbr',
 'admission_type_id',
 'discharge_disposition_id',
 'admission_source_id',
 'time_in_hospital',
 'num_lab_procedures',
 'num_procedures',
 'num_medications',
 'number_outpatient',
 'number_emergency',
 'number_inpatient',
 'number_diagnoses']

In [38]:
df[numeric_columns].describe().T

,count,mean,std,min,25%,50%,75%,max
encounter_id,101766.0,1.652016e+08,1.026403e+08,12522.0,84961194.0,152388987.0,2.302709e+08,443867222.0
patient_nbr,101766.0,5.433040e+07,3.869636e+07,135.0,23413221.0,45505143.0,8.754595e+07,189502619.0
admission_type_id,101766.0,2.024006e+00,1.445403e+00,1.0,1.0,1.0,3.000000e+00,8.0
discharge_disposition_id,101766.0,3.715642e+00,5.280166e+00,1.0,1.0,1.0,4.000000e+00,28.0
admission_source_id,101766.0,5.754437e+00,4.064081e+00,1.0,1.0,7.0,7.000000e+00,25.0
time_in_hospital,101766.0,4.395987e+00,2.985108e+00,1.0,2.0,4.0,6.000000e+00,14.0
num_lab_procedures,101766.0,4.309564e+01,1.967436e+01,1.0,31.0,44.0,5.700000e+01,132.0
num_procedures,101766.0,1.339730e+00,1.705807e+00,0.0,0.0,1.0,2.000000e+00,6.0
num_medications,101766.0,1.602184e+01,8.127566e+00,1.0,10.0,15.0,2.000000e+01,81.0
number_outpatient,101766.0,3.693572e-01,1.267265e+00,0.0,0.0,0.0,0.000000e+00,42.0
